pkg/oauth/tokenrequest/basicauth.go

```diff
@@ -0,0 +1,160 @@
+package tokenrequest
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+// BasicAuthNoUsernameError when basic authentication challenge handling was attempted
+// but the required username was not provided from the command line options
+type BasicAuthNoUsernameError struct{}
+
+//
+func (e *BasicAuthNoUsernameError) Error() string {
+	return "did not receive username. Pass 'oc login --username <username>' for the password prompt"
+}
+
+// NewBasicAuthNoUsernameError returns an error for a basic challenge without a username
+func NewBasicAuthNoUsernameError() error {
+	return &BasicAuthNoUsernameError{}
+}
+
+func NewBasicChallengeHandler(
+	host string,
+	reader io.Reader,
+	writer io.Writer,
+	passwordPrompter PasswordPrompter,
+	username, password string,
+) *BasicChallengeHandler {
+	return &BasicChallengeHandler{
+		Host:             host,
+		Reader:           reader,
+		Writer:           writer,
+		passwordPrompter: passwordPrompter,
+		Username:         username,
+		Password:         password,
+	}
+}
+
+type BasicChallengeHandler struct {
+	// Host is the server being authenticated to. Used only for displaying messages when prompting for username/password
+	Host string
+
+	// Reader is used to prompt for username/password. If nil, no prompting is done
+	Reader io.Reader
+	// Writer is used to output prompts. If nil, stdout is used
+	Writer io.Writer
+
+	// passwordPrompter is used to retrieve the password as a string
+	passwordPrompter PasswordPrompter
+
+	// Username is the username to use when challenged. If empty, a prompt is issued to a non-nil Reader
+	Username string
+	// Password is the password to use when challenged. If empty, a prompt is issued to a non-nil Reader
+	Password string
+
+	// handled tracks whether this handler has already handled a challenge.
+	handled bool
+	// prompted tracks whether this handler has already prompted for a username and/or password.
+	prompted bool
+}
+
+func (c *BasicChallengeHandler) CanHandle(headers http.Header) bool {
+	isBasic, _ := basicRealm(headers)
+	return isBasic
+}
+func (c *BasicChallengeHandler) HandleChallenge(requestURL string, headers http.Header) (http.Header, bool, error) {
+	if c.prompted {
```

Comment:
please add a note in godoc that this is not guarded, hence unsafe for concurrent use.